reference implementation of the Sensor Markup Language (SenML)

backenklee opened the pull request:
This implementation supports only JSON for now, but support for CBOR is possible as well and I'm planning to implement it soonish. Data can be retrieved from a SenML document in JSON format and values will be stored in a struct (senml_record_t) as normal C types (char *, int, double etc). Encoding such a struct is also possible and will create an according SenML document.

See https://tools.ietf.org/html/draft-jennings-core-senml-06

bapclenet commented:
Hi, 
Nice implementation, do you have an exemple that I can try while reviewing the source code?

gebart commented:
PS. There is a new version of the standard which can be found at https://tools.ietf.org/html/draft-ietf-core-senml-00

bapclenet commented:
@backenklee, does your implementation implement the modification added in senml-00?

gebart commented:
For reference: It looks like the biggest modification in the spec between jennings-core-senml-06 to ietf-core-senml-00 was to change the floating point values into double precision.

backenklee commented:
@gebart thanks for pointing me to that, looks like the WG finally adopted the draft :)

@bapclenet my implementation is based on jennings-core-senml-06, since i wasn't aware of core-senml-00 until today. however, with regard to @gebart 's comment, i have been using double for floating point values all along.

here's an example:

https://gist.github.com/backenklee/aac242f0ef1950cb49ef7c8e547d994a

bapclenet commented:
@backenklee, ok so no need to port to new implementation if you already use double. 
Thanks for the example, I'll try it

gebart commented:
An interesting optimization in a CBOR implementation would be to use the data type that gives the same precision in the minimum amount of bytes, e.g. use different width ints if the value is integer and less than `INT*_MAX` and to use float if the extra bits in the double representation are all zeroes, but that is unrelated to this PR since this does not implement CBOR.

bapclenet commented:
@gebart, I let you review since you seem to better know than me what SenML is

backenklee commented:
The link to the draft in the documentation now points to the most recent version. The only noteworthy change is indeed to switch from float to double (see https://tools.ietf.org/html/draft-ietf-core-senml-00 ), which my implementation uses already.

@gebart For clarification, are you talking about encoding or decoding? For encoding I completely agree. But if you meant decoding, it would make the implementation unnecessarily complicated, wouldn't it?

gebart commented:
@backenklee I was talking about an encoder to minimize the number of bytes sent over the air in order to conserve power.

backenklee commented:
The implementation is now fully compliant to the draft (including handling of binary data which was missing before), and includes an example application.

Yonezawa-T2 commented on sys/senml/senml.c:
```diff
@@ -0,0 +1,328 @@
+
+    jsmn_init(&parser);
+
+    jsmn_parse(&parser, input, strlen(input), tokens, num_toks);
```
Missing reassignment to `num_toks`? Or the following check is redundant.

backenklee replied:
You're right, fixed it.

Yonezawa-T2 commented on sys/senml/senml.c:
```diff
@@ -0,0 +1,328 @@
+                else if (!strcmp(tmp_key, SJ_BASE_VALUE)) {
+                    pack->base_info->base_value = atof(tmp_val);
+                }
+                else {   // no more base information
```
Because [“An object is an unordered set of name/value pairs.”](http://www.json.org/), the follwoing should be valid:

```json
[
  { "t": 1276020076, "v":23.5, "u":"Cel", "bn": "urn:dev:ow:10e2073a01080063" }
  { "t": 1276020091, "v":23.6, "u":"Cel" }
]
```

Yonezawa-T2 commented on sys/senml/senml.c:
```diff
@@ -0,0 +1,328 @@
+            tmp_val = &input[t->start];
+            input[t->end] = '\0';
+
+            if (array_num == 0 && !base_values_checked) {
```
Because [“An object is an unordered set of name/value pairs.”](http://www.json.org/), the follwoing should be valid:

```json
[
  { "t": 1276020076, "v":23.5, "u":"Cel", "bn": "urn:dev:ow:10e2073a01080063" }
  { "t": 1276020091, "v":23.6, "u":"Cel" }
]
```

Therefore, `&& !base_values_checked` should be moved to `if (tokens[i + 1].type == JSMN_OBJECT` line and `base_values_checked` should be renamed to better name.

Yonezawa-T2 commented on sys/senml/senml.c:
```diff
@@ -0,0 +1,328 @@
+                    base_values_checked = true;
+                }
+
+                if (tokens[i + 1].type == JSMN_OBJECT) {
```
If data has only one record, this may access outside of the token array.

Yonezawa-T2 commented on sys/senml/senml.c:
```diff
@@ -0,0 +1,328 @@
+        }
+
+        if (t->type == JSMN_OBJECT) {
+            array_num++;
```
Should we check `array_num < pack->num` after this line?

backenklee replied:
Yes, we should :) Check is included now.

Yonezawa-T2 commented on sys/include/senml.h:
```diff
@@ -0,0 +1,161 @@
+    char              *name;        //!< Sensor's name, will be appended to base name if provided
+    char              *unit;        //!< Unit of the measurement
+    double             time;        //!< Time (seconds since unix epoch) of the measurement
+    unsigned int       update_time; //!< Time before sensor provides an updated measurement
```
Update time should be double (see [Section 8](https://tools.ietf.org/html/draft-ietf-core-senml-00#section-8)).

backenklee replied:
Thanks for pointing this out, is fixed.

backenklee commented:
Hi @Yonezawa-T2 , thank you very much for your review! I agree with all your comments and fixed the issues accordingly, with one exception:

The working group agreed that records without values (i.e. containing base information only) should not be allowed. The draft will be updated accordingly soon, see

https://mailarchive.ietf.org/arch/msg/t2trg/42RXIPea-20yexOXmEzPO_45feU

and

https://tools.ietf.org/wg/core/minutes?item=minutes-95-core.html

So I just reverted the commits that allowed this.

Yonezawa-T2 commented on sys/senml/senml.c:
```diff
@@ -0,0 +1,325 @@
+int senml_encode_json_s(const senml_pack_t *pack, char *output, size_t len)
+{
+    size_t insert_pos = 0;
+
```
For foolproof, we should ensure `len >= 3`.

backenklee replied:
Yes, we should

Yonezawa-T2 commented on sys/senml/senml.c:
```diff
@@ -0,0 +1,325 @@
+    insert_pos++;
+    len--;
+
+    // using len - 2 to leave space for closing } and ]
```
As `size_t` is unsigned, `len` may underflow if `len < 2`, results in buffer overrun.

Yonezawa-T2 replied:
We need 3 bytes for }, ], and '\0'.

backenklee replied:
True and true, both fixed

Yonezawa-T2 commented on sys/senml/senml.c:
```diff
@@ -0,0 +1,325 @@
+    for (size_t i = 0; i < pack->num; i++) {
+        senml_record_t *curr_record = &(pack->records[i]);
+
+        if (i != 0) {
```
If we don't have a `base_info`, the first '{' is not inserted.

backenklee replied:
Good spotting, also fixed

Yonezawa-T2 commented on sys/senml/senml.c:
```diff
@@ -0,0 +1,325 @@
+        senml_record_t *curr_record = &(pack->records[i]);
+
+        if (i != 0) {
+            output[insert_pos] = '{';
```
This is not safe if `len` is 0. For example:

`output`: `"[{\"v\":1},"`
`insert_pos`: 9
`len`: 0

backenklee replied:
Should not be an issue anymore, since the function will return -1 if cnt_len would be 0 after a call to snprintf

Yonezawa-T2 commented on sys/senml/senml.c:
```diff
@@ -0,0 +1,325 @@
+                                         curr_record->value.f);
+            }
+            else if (curr_record->value_type == SENML_TYPE_STRING) {
+                chars_written = snprintf(&output[insert_pos], len - 2, "\"vs\":\"%s\"",
```
Special characters (i.e. ", \, and control-characters) should be escaped.

backenklee replied:
True, I'm still thinking about how to do this nicely and without wasting too much memory...

backenklee commented:
@Yonezawa-T2 again, thank you very much for the review! I apologize for the late reaction. I fixed all issues except the escaping of special characters, see comment above. Would be nice if RIOT provided a JSON generator, because such a function would belong there, but I'll figure something out.

Yonezawa-T2 commented on sys/senml/senml.c:
```diff
@@ -0,0 +1,357 @@
+    cnt_len--;
+
+    if (pack->base_info->version) {
+        chars_written = snprintf(&output[insert_pos], cnt_len, "\"bver\":%u,",
```
If `len == 4`, `cnt_len == -1` here. `cnt_len` will be casted to `size_t` because `snprintf` receives `size_t`, results in underflow.

Yonezawa-T2 commented on sys/senml/senml.c:
```diff
@@ -0,0 +1,357 @@
+        }
+
+        if (curr_record->name) {
+            chars_written = snprintf(&output[insert_pos], cnt_len, "\"n\":\"%s\",",
```
`cnt_len` may be zero at the end of the loop, so it will be decremented to `-1` in the next iteration after inserting `{`.

Abstract execution trace of `cnt_len`:

```C
if (curr_record->name) {
  chars_written = snprintf(&output[insert_pos], cnt_len, "\"n\":\"%s\",",
                           curr_record->name);

  if (chars_written >= cnt_len || chars_written < 0) {
    return -1;
  }

  insert_pos += chars_written;
  cnt_len -= chars_written;

  //////////////////
  // cnt_len >= 1 //
  //////////////////
 }

// ...

// decrease insert_pos to overwrite trailing comma
insert_pos--;
cnt_len++;
//////////////////
// cnt_len >= 2 //
//////////////////

output[insert_pos] = '}';
insert_pos++;
cnt_len--;
//////////////////
// cnt_len >= 1 //
//////////////////

output[insert_pos] = ',';
insert_pos++;
cnt_len--;
//////////////////
// cnt_len >= 0 //
//////////////////

/////////////////////
// next iteratrion //
/////////////////////

i++;

senml_record_t *curr_record = &(pack->records[i]);

if (i != 0) {
  output[insert_pos] = '{';
  insert_pos++;
  cnt_len--;
 }

///////////////////
// cnt_len >= -1 //
///////////////////

if (curr_record->name) {
    /////////////////////////////////
    // cnt_len is casted to size_t //
    //         UNDERFLOW!          //
    /////////////////////////////////
    chars_written = snprintf(&output[insert_pos], cnt_len, "\"n\":\"%s\",",
                             curr_record->name); 

/// ...

```

Yonezawa-T2 commented on sys/senml/senml.c:
```diff
@@ -0,0 +1,357 @@
+    insert_pos++;
+
+    output[insert_pos] = '\0';
+    insert_pos++;
```
`ssize_t cnt_len = len - 3;` can be `ssize_t cnt_len = len - 1;`.

Invariant:
`insert_pos + cnt_len == len - 3`
so
`insert_pos == len - 3 - cnt_len`.

Abstract execution trace:

```C
    // ...

    output[insert_pos] = ',';
    insert_pos++;
    cnt_len--;
    ///////////////////////////
    // cnt_len >= 0          //
    // insert_pos <= len - 3 //
    ///////////////////////////
}

// remove trailing comma or unnecessary opening curly brace
if (insert_pos >= 2) {
    insert_pos--;
    ///////////////////////////   
    // insert_pos <= len - 4 //
    ///////////////////////////
}

output[insert_pos] = ']';
insert_pos++;
///////////////////////////   
// insert_pos <= len - 3 //
///////////////////////////

output[insert_pos] = '\0';
insert_pos++;
///////////////////////////   
// insert_pos <= len - 2 //
///////////////////////////
```

`insert_pos` should be `<= len` here instead of `<= len - 2`, so initial value of `cnt_len` should be `len - 1` rather than `len - 3`.

backenklee commented:
I think I managed to fix the over-/underflow issues this time without introducing new ones, and I hope we can ship this feature with the new release :)

gebart commented:
I haven't done a proper review of the code, but two things I noticed that we need to fix:

 - The example in `examples/senml_json` is more like the other tiny test applications in `tests/`, you should move it.
 - Unit tests would be very useful to avoid regressions in the future.

backenklee commented:
I have added some simple tests that en- and decode different documents. I think the example is still useful because it shows the basic features and is probably where most people who want to use this library will look. So I simplified it a little. What do you think?

Yonezawa-T2 commented on sys/senml/senml.c:
```diff
@@ -0,0 +1,416 @@
+            buf_len -= chars_written;
+        }
+
+        if (buf_len == buf_len_old) {   // record was empty, no values inserted: remove '{'
```
If `base_info` is available and this is the first record, '{' is not inserted.

Yonezawa-T2 replied:
Should we consider empty record? As at least value or sum is required, empty record is invalid.
https://tools.ietf.org/html/draft-ietf-core-senml-02#section-4.2

backenklee replied:
No, the '{' does get inserted, check [line 274 in senml.c](https://github.com/RIOT-OS/RIOT/pull/5544/files#diff-17dca0f32589a71ffb5cf494b254ecdaR274), and the test application. It includes documents without base info.

The check for empty records was included with the following scenario in mind:
Suppose you initialize an array of `senml_record_t` of length 4, but then (for some reason) you can only fill 2 or 3 of the records. `pack.num` will still be 4, but the resulting JSON document will only contain records with actual content (the '{' for the empty records will be removed).

Yonezawa-T2 replied:
I mean, '{' is not inserted at line 275 (it is inserted at line 178 before inserting base infos), so that we should not delete char before cursor, that is ','.

backenklee replied:
Ah, I see, you mean when the first records is empty, right? I added a boolean variable to check for this edge case, output is now correct when the first, the last, or a record in between is empty.

Yonezawa-T2 replied:
`senml_decode_json_s` sets `pack->num` to actual number of valid records. So `pack->num` may differ from the number of allocated records. Therefore, requiring caller of `senml_encode_json_s` to set `pack->num` to actual number of valid records may be acceptable.

Ignoring empty records my be less useful than expected. Typical application sends data repeatedly to a server and reuses a statically allocated pack. Caller must clear records before reusing the pack to make `senml_encode_json_s` ignore empty records correctly. Setting `pack->num` before calling `senml_encode_json_s` is easier than clearing the pack before reusing it. 

Example:
Suppose a RIOT node with a sensor device interrupting at irregular intervals. The node sends data to a server on timer, or when no space available in the buffer.
```C
#define N 10

static senml_record_t records[N]; 
static senml_pack_t pack;
static int next_record = 0;

static void init_pack() {
  pack.records = records;
  ... // init other fields and records.
}

void on_new_sensor_data_available(double time, double data) {
  if (next_record == N) {
    send_seml_to_server();
  }
  pack.records[next_record].time = time;
  pack.records[next_record].value.f = data;
  next_record++;
}

void on_timer() {
  send_seml_to_server();
}

void send_seml_to_server() {
  if (next_record == 0) {
    return;
  }
  pack.num = next_record;

  char buffer[1024];

  senml_encode_json_s(&pack, buffer, 1024);
  send_string_to_server(buffer);
  next_record = 0;
  // does not need to clear records here
}
```

backenklee replied:
I agree about the client/server use case, but what about a network of devices that act as clients and servers simultaneously to exchange data with each other? If a programmer makes a small mistake or misses an edge case (like I did in this implementation), it would be very frustrating if an application stops working because `senml_encode_json_s` produces invalid documents

Yonezawa-T2 replied:
I see. Let's skip empty records.

Yonezawa-T2 commented on sys/senml/senml.c:
```diff
@@ -0,0 +1,416 @@
+            buf_len -= chars_written;
+        }
+
+        if (curr_record->value_sum != 0) {
```
Does 0 mean no sum?

backenklee replied:
For lack of a better solution, yes. Can you think of a situation where this would be problematic and could not be solved/clarified by previous or subsequent records?

Yonezawa-T2 replied:
NaN would be a better marker to indicate no sum because JSON does not allow NaN. Correctness is more important than efficiency for reference implementation. Maybe we should provide a clearing function like this:

```C
senml_record_t seml_empty_record = {
  .name = NULL,
  .unit = NULL,
  .time = NAN,
  .update_time = NAN,
  .value_sum = NAN,
  .value_type = SENML_TYPE_UNDEF,
};

inline void seml_clear_record(senml_record_t *record) {
  *record = seml_empty_record;
} 
```

backenklee replied:
NaN would be nice, but it is safe to use on all platforms? This would also add complexity to the creation of empty records, because `memset(&record, 0, sizeof(record))` would not be sufficient anymore

Yonezawa-T2 replied:
You are right. C99 doesn't seems to mandate NaN support (e.g. 7.12.11.2).

Yonezawa-T2 commented on sys/senml/senml.c:
```diff
@@ -0,0 +1,427 @@
+
+        size_t buf_len_old;
+
+        if ((i != 0 || ins_pos == 1) && insert_opening_brace) {
```
We can drop `(i != 0 || ins_pos == 1) &&` if we insert `insert_opening_brace = false` after 183.

Yonezawa-T2 commented on sys/senml/senml.c:
```diff
@@ -0,0 +1,427 @@
+
+        if (buf_len == buf_len_old) {   // record was empty, no values inserted: remove '{'
+            if (i != 0) {
+                ins_pos--;
```
What if `base_inf` exists and 1st and 2nd record are empty? More straight method would be checking `output[ins_pos - 1] == '{'`.

kYc0o commented:
Since this is a new feature that is still in review, I'm forced to mark it for the next one.

haukepetersen commented:
@backenklee any new on this?

miri64 commented:
Postponed due to Feature Freeze.

OlegHahm commented:
@backenklee, ping!

backenklee commented:
I implemented @Yonezawa-T2 's suggestions and added according tests. The output is valid JSON and valid SenML.

smlng commented:
@backenklee what are the major differences between the example and the test? They look kind of redundant, if so, one should do?!

backenklee commented:
The example briefly shows how the implementation is supposed to be used; the tests check if it still produces correct output when *not* used as intended. The example should definitely stay (the examples folder is also where most people look first when they want to know how to use a library), but if this implementation receives positive reviews, I can remove the tests.

miri64 commented:
> The example should definitely stay (the examples folder is also where most people look first when they want to know how to use a library), but if this implementation receives positive reviews, I can remove the tests.

I think they should both stay, but maybe you can integrate the tests into the unittests?

miri64 commented on sys/senml/senml.c:
```diff
@@ -0,0 +1,429 @@
+    buf_len--;
+
+    if (!pack->base_info) {
+        goto encode_records;
```
Oo?

miri64 replied:
I mean putting everything between this and the `encode_records` label into an `if (pack->base_info)` branch would make the code far more readable.

backenklee replied:
Agreed; changed accordingly. My intention was to reduce indentations (to make the code more readable, haha...), but I see your point.

backenklee commented:
> I think they should both stay, but maybe you can integrate the tests into the unittests?

Unit tests seem like a good idea, but how would I handle the floating point numbers? For example: If the input string contains the value 1.29, then after decoding it, the struct may contain something like 1.2899999999999. Then after encoding it again, printf might turn this into 1.28999. This is all correct behavior, but an TEST_ASSERT_EQUAL_... would fail. Do you have any suggestions on how to handle that?

gebart commented:
@backenklee for floating point numbers you should compute the difference between the two numbers and compare the absolute value to a small threshold (epsilon)

see for example http://www.cygnus-software.com/papers/comparingfloats/Comparing%20floating%20point%20numbers.htm

miri64 commented:
@backenklee what @gebart is saying ;-)

backenklee commented:
@gebart @miri64 great resource, thank you! as you can see, I removed the old tests, implemented unit tests, and brought the implementation up to speed with the latest version of the draft (supports now base sum and links).

miri64 commented on tests/unittests/tests-senml_json/tests-senml_json.c:
```diff
@@ -0,0 +1,563 @@
+
+char output[512];
+
+bool double_almost_equal(float a, float b)
```
static

backenklee replied:
whoopsie, that one slipped through

miri64 commented on tests/unittests/tests-senml_json/tests-senml_json.c:
```diff
@@ -0,0 +1,563 @@
+                                "\\\"bar\\\"}]\",\"t\":1480778629.230000,\"ut\":120.000000,"
+                                "\"vd\":\"VGhpcyBpcyBiaW5hcnkgdmFsdWUK\"}]";
+
+char output[512];
```
Actually those can be `static`, too, the ones above even `static const`?

backenklee replied:
`static`, yes, but they can't be `const` because the `char *`s in a record point to a position in the input string into which `senml_decode_json_s` inserts `'\0'` to separate the string values

backenklee replied:
`static`, yes, but not `const`. `senml_decode_json_s` inserts `'\0'` after a string value and a `char *` in a record will point to a position in the input string. Otherwise, strings would have to be copied, and I felt like that would be too ugly without malloc and waste memory.

backenklee replied:
ooops, sorry for double posting, github was acting weird...

miri64 commented on sys/include/senml.h:
```diff
@@ -0,0 +1,167 @@
+#endif  // __cplusplus
+
+
+#define SENML_SUPPORTED_VERSION (5)    //!< The highest SenML version this implementation supports
```
Here and in other places: Please use C-style comments `/* */` (`/**< */` for inline doxygen ;-))

miri64 commented on sys/include/senml.h:
```diff
@@ -0,0 +1,167 @@
+    char               *base_name;       //!< Concatenated with name attribute gives sensor's UID
+    double              base_time;       //!< Update time values are relative to this time
+    char               *base_unit;       //!< When all values have the same unit, set this field
+    double              base_value;
```
Doc missing

miri64 commented on sys/include/senml.h:
```diff
@@ -0,0 +1,167 @@
+} senml_value_type_t;
+
+
+/*! struct that contains base information which applies to all subsequent entries */
```
Please use

```C
/**
 * @brief
 */
```

for top-doc

miri64 commented on sys/include/senml.h:
```diff
@@ -0,0 +1,167 @@
+ * @param[out] pack  The <code>senml_pack_t</code> structure in which the
+ *                   records and base info will be stored.
+ *
+ * @return 0 on success, -1 if \p input is not a valid SenML document, or a
```
Please don't mix `\` and `@` style doc. (use `@`)

miri64 commented on sys/include/senml.h:
```diff
@@ -0,0 +1,167 @@
+ * @brief Decodes a SenML document in JSON format
+ *
+ * @param[in]  input The JSON document containing the SenML pack
+ * @param[out] pack  The <code>senml_pack_t</code> structure in which the
```
Use `@ref senml_pack_t` (also markdown doc formatting is preferred ;-))

miri64 commented on sys/senml/senml.c:
```diff
@@ -0,0 +1,456 @@
+            last_type = JSMN_OBJECT;
+            array_num++;
+
+            // the cast to unsigned int is ok because array_num will be >= 0 at this point
```
C-style comments, please

miri64 commented on sys/senml/senml.c:
```diff
@@ -0,0 +1,456 @@
+
+int senml_encode_json_s(const senml_pack_t *pack, char *output, size_t len)
+{
+    if (len < 3) {           // we need space for at least '[', ']', and '\0'
```
C-style comments, please (also below)

backenklee commented:
Thank you @miri64 for taking the time :-) Does anyone else have any issues or an improvement they would like to see?

backenklee commented:
Hi everyone, I know you have a lot on your plate, but it is currently unclear to me what is missing before this PR can be merged. I implemented all the requested changes (at least looks that way to me) and the unit tests.

Also, the latest commit (only small changes in the docs) seems to have broken jenkins (?), and I can't figure out why. Is this important? Is it my fault?

I would love to see this implementation included in the upcoming release, please let me know what I have to do to make that happen.

OlegHahm commented:
Sorry for the delay - I think this one slipped through somehow. At a first glance, I think this thing should be ready to be merged, but maybe I miss something. @miri64?

PeterKietzmann commented:
@backenklee planned feature freeze is today. However, **maybe** we will postpone the freeze until Monday. In that case you would need to address all comments and find someone who ACKs this PR. But in case it won't make it in this release, it's worth it to start working on it so it will be part of the further next relase (which sometimes comes faster than expected)

PeterKietzmann commented:
Ah, Oleg was 13 seconds faster than me :-)

kaspar030 commented on sys/include/senml.h:
```diff
@@ -0,0 +1,171 @@
+
+
+/**
+ * @ingroup     senml
```
is this group defined somewhere?

miri64 replied:
Also, please prefix with `sys_`

miri64 replied:
Also, if you want to add it to an existing group its `@addtogroup`, `@ingroup` makes it a subgroup of `senml`. I think you want to use `@defgroup` here however and use `sys` as your `@ingroup`. Remember to also provide a human-readable name for the `senml` group. I propose `SenML`

backenklee replied:
oopsy, not yet. is it ok if i just put

`@defgroup senml Sensor Measurement List`

before the ingroup command?

miri64 replied:
Might be wrong, but doesn't SenML stands for Sensor Markup Language?

kaspar030 commented on sys/include/senml.h:
```diff
@@ -0,0 +1,171 @@
+ *
+ * @author      Lennart Dührsen <lennart.duehrsen@fu-berlin.de>
+ */
+
```
please use single empty lines here and throughout the file

kaspar030 commented on sys/senml/Makefile:
```diff
@@ -0,0 +1,3 @@
+MODULE = senml
```
This line is redundant if the folder name is "senml"

kaspar030 commented on tests/unittests/tests-senml_json/tests-senml_json.h:
```diff
@@ -0,0 +1,38 @@
+#ifdef __cplusplus
+extern "C" {
+#endif
+
```
single empty lines

miri64 commented on sys/include/senml.h:
```diff
@@ -0,0 +1,171 @@
+ *         positive number indicating by how much the number of records in
+ *         @p input exceeds `pack->num`
+ */
+int senml_decode_json_s(char *input, senml_pack_t *pack);
```
From doc this should probably be `const char *input`

backenklee replied:
it can't, see https://github.com/RIOT-OS/RIOT/pull/5544#discussion_r90777588

miri64 replied:
Sorry, forgot about this :(

miri64 commented on tests/unittests/tests-senml_json/tests-senml_json.c:
```diff
@@ -0,0 +1,562 @@
+
+#include "tests-senml_json.h"
+
+static char senml_json_empty_doc[]   = "[]";
```
If https://github.com/RIOT-OS/RIOT/pull/5544/files#r97090882 is addressed these strings probably all can be `const`.

backenklee replied:
see above

miri64 commented:
(please also address @kaspar030's comments. @kaspar030 if you request changes, please mark your review as such)

kaspar030 commented:
> @kaspar030 if you request changes, please mark your review as such)

Yeah, you know, you just want to make a small comment, then scroll down, more comes up...

backenklee commented:
It used to, they changed it because $reasons...

backenklee commented:
@kaspar030 hm, should i add a BOARD_INSUFFICIENT_MEMORY list to the makefile? and is there a blacklist command, or how would i deal with missing atof support?

miri64 commented:
I think blacklisting these boards for now seems like the quickest way to go.

backenklee commented:
ok, but just for the unittests or the whole thing?

miri64 commented:
For any application that uses the module. You can't blacklist boards for a module.

OlegHahm commented:
Wait a second.

OlegHahm commented:
> all msp430 fail because of missing atof

Why is atof() missing for MSP430? In m y toolchain it is available and that is pretty old:
`msp430-gcc (mspgcc_20120406) 4.6.3 20120301 (mspgcc LTS 20120406 unpatched)`

OlegHahm commented:
In the case that there are indeed toolchains that do not provide an atof() implementation I would vote to provide this function ourselves - if this is the only thing that is missing.

backenklee commented:
@OlegHahm thank you for taking the time to look at this! i really have no idea what's going on the build server.

what do you propose with regard to this PR? can we merge it, and remove the blacklisted boards as soon as we have an atof() implementation of our own?

OlegHahm commented:
@backenklee, I would propose to check with @kaspar030 why building is failing on MSP430 platforms in the CI. If the problem cannot be solved by updating the toolchain, we could go with our own implementation.

backenklee commented:
Hey @kaspar030 , could it be that the build serve still uses the toolchain recommended in the [wiki](https://github.com/RIOT-OS/RIOT/wiki/Family%3A-MSP430) ? That one does indeed not provide an atof() implementation, see https://sourceforge.net/p/mspgcc/msp430-libc/ci/master/tree/src/stdlib/ . It is also obsolete and superseeded by http://www.ti.com/tool/msp430-gcc-opensource . Would check if that is the case here?

miri64 commented:
afaik Murdock is using the docker image provided at https://github.com/RIOT-OS/riotdocker/. Which uses the [`gcc-msp430` provided with Ubuntu](https://github.com/RIOT-OS/riotdocker/blob/master/Dockerfile#L71-L72)

backenklee commented:
thanks @miri64 , that is indeed mspgcc packaged for ubuntu, 17.04 seems to use mspgcc-20120406 (see https://launchpad.net/mspgcc/+packages ). i just grep'ed through the files of the mspgcc-msp430-libc, and i cannot find an atof() implementation (atoi, atol are there). @OlegHahm since you use that version, are you positive that it provides atof()..? otherwise, that would explain why the builds fail.

side note: i also seem to be the [first person](https://github.com/RIOT-OS/RIOT/search?utf8=%E2%9C%93&q=atof) to contribute code to RIOT that makes use of atof() o.o

backenklee commented:
there's also a project called [Energia](http://energia.nu/), whose goal it is to make arduino software run on MSP430 platforms (among others). they use mspgcc too, and also found that it does not provide atof. the [code](https://github.com/energia/Energia/commit/87cdb3e22275ba60de936070a8479f0c6a7bbac4) is available under GPL, maybe we can use their implementation (and/or switch toolchains)?

miri64 commented:
Any plans do progress on this?

aabadie commented:
ping ! Do you think you can make it for the next release (4 weeks left)

cgundogan commented:
@backenklee I retriggered murdock to have an overview about what fails and what succeeds with the current setup. I propose that you squash the commits and from there I will do a quick round of reviewing. If nothing serious pops up we can go ahead and merge this finally ..

cgundogan commented:
that seems to be easy:
1. static tests fail; this should be easy to fix
2. unittests for msba2 fails => blacklist

miri64 commented:
> 2. unittests for msba2 fails => blacklist

*raises eyebrow*

cgundogan commented:
>raises eyebrow

well that's our current best way to solve memory issues until someone splits the unittests into separate binaries.

miri64 commented:
> well that's our current best way to solve memory issues until someone splits the unittests into separate binaries.

Oh, because you said "blacklist" I assumed the tests to fail due to some stuff related to the platform, not because of the image's size. Blacklisting and adding the board to the list of boards with insufficient memory are two very different approaches. The first means the board isn't build at all for the application (which I think is a very bad idea for an application with such a huge coverage), the latter that just the linking step (which causes the overflow error) is skipped.

emmanuelsearch commented:
@backenklee so what's the status here?

emmanuelsearch commented:
@kaspar030 are you still requiring changes to this PR?

haukepetersen commented on sys/include/senml.h:
```diff
@@ -0,0 +1,158 @@
+    double             value_sum;   /**< Integrated sum of the values over time                   */
+    senml_value_type_t value_type;  /**< Indicates which type the value is of                     */
+    union {
+        double            f;  /**< A float value                              */
```
are you sure you want to use `double` (64-bit long!) for handling floating point numbers? This seems to be very bad for contrained devices in multiple ways:
- a lot o memory is needed to store them
- even most HW floating point units do not support them (32-bit float only)

haukepetersen commented on sys/include/senml.h:
```diff
@@ -0,0 +1,158 @@
+#define SC_LINK          (9)    /**< Key for the link attribute in CBOR documents          */
+
+/**
+ * @brief The different data types SenML supports
```
Check for this file: sometime you start the comments with a capital letter, sometimes with small letter.

You might also want to put 3 space chars after the @brief (tab alignment) - this is the default in RIOT...

haukepetersen commented on sys/senml/senml.c:
```diff
@@ -0,0 +1,467 @@
+
+#include "senml.h"
+#include "jsmn.h"
+#include "cbor.h"
```
nothing from this file is currently used in the implementation, right? So just drop the include for now.

haukepetersen commented on sys/include/senml.h:
```diff
@@ -0,0 +1,158 @@
+ * @param[out] output The resulting JSON document
+ * @param[in]  len    The length of @p output in bytes
+ *
+ * @return 0 on success, or -1 when @p output is not big enough
```
I would prefer to add named return values, something like
```c
// somewhere in the top of the file
enum {
    SENML_OK     = 0,
    SENML_OVRFLW = -1,
    ...
};

// here
/**
 * @return  SENML_OK on success
 * @return  SENML_OVRFLW if @p output is not large enough
 */
```
Same goes of course for `senml_decode_json_s()`

haukepetersen commented on sys/include/senml.h:
```diff
@@ -0,0 +1,158 @@
+ * @brief struct that holds a SenML pack (optional base info and 1..n records)
+ */
+typedef struct {
+    senml_base_info_t  *base_info;  /**< Pointer to the base info, may be NULL           */
```
What is the reason for having `senml_base_info_t` as separate type and not merge it directly into the `senml_pack_t` structure? In the implementation it is used always through the `senml_pack_t` structure anyway... IMHO this would not only simplify the interface, but also save some de-referencing operations...

haukepetersen commented on tests/unittests/tests-senml_json/tests-senml_json.c:
```diff
@@ -0,0 +1,562 @@
+                                "\"vd\":\"VGhpcyBpcyBiaW5hcnkgdmFsdWUK\"}]";
+
+
+static bool double_almost_equal(float a, float b)
```
In the senml implementation you use `double`, here `float`. Any reason for this?

haukepetersen commented:
Besides a couple of style issues, I see two major issues with this PR that I'd like to see addressed:
- the use of `double` is IMHO very questionable for constrained platforms (where even float is sub-optimal)
- and even more pressing: the code does not work for Cortex-M (and possibly other platforms, tested for `samr21-xpro`): the example application hard-faults:
```
...
2017-12-19 12:22:55,084 - INFO # Data Value:	U2VuTUwgaXMgc28gY29vbCEK
2017-12-19 12:22:55,084 - INFO # 
2017-12-19 12:22:55,085 - INFO # SenML is so cool!
2017-12-19 12:22:55,086 - INFO # 
2017-12-19 12:22:55,091 - INFO # [{"bver":5,"bn":"urn:dev:mac:0024befffe804ff2;","bt":,
2017-12-19 12:22:55,105 - INFO # main(): This is RIOT! (Version: 2018.01-devel-1092-g43159-Prag-HEAD)
2017-12-19 12:22:55,107 - INFO # Version:	5
2017-12-19 12:22:55,110 - INFO # Base Name:	urn:dev:mac:0024befffe804ff1
2017-12-19 12:22:55,111 - INFO # Base Time:	
2017-12-19 12:22:55,113 - INFO # Base Unit:	NULL
2017-12-19 12:22:55,114 - INFO # Base Value:	
2017-12-19 12:22:55,114 - INFO # 
...```
I am pretty sure this is related to the use of `float/double` with `printf`...

smlng commented:
postponed

tcschmidt commented:
Is this still worked on or can this be just closed?

tcschmidt commented:
Closed due to inactive author.
